product_configurator/models/product.py

```
@@ -466,11 +469,12 @@ def validate_configuration(self, value_ids, custom_vals=None, final=True):
                 common_vals = set(value_ids) & set(line.value_ids.ids)
                 custom_val = custom_vals.get(attr.id)
                 if line.required and not common_vals and not custom_val:
-                    # TODO: Verify custom value type to be correct
-                    return False
+                    # If there are possible values, then enforce
+                    # the "required" status.
+                    if set(line.value_ids.ids) & set(avail_val_ids):
+                        # TODO: Verify custom value type to be correct
+                        return False
```

I do agree that there might be a case where a user might want "required if options available" rather than flat required.

On one hand you can have this flexibility on the other hand you are not guaranteed to have a required value and if you mess up your configuration restrictions the configuration will silently pass.

Maybe there's a way we can differentiate between two cases or also provide an error message when a line is required and has certain configurations which do not provide any options for the required line.

What do you think?

---

@PCatinean My initial feeling was similar, and I was going to make a 2 way flag.  But I fell back to this - if there are no possible selections based on other domains, then it just seemed logical that it can't be invalid - the definition must be...

If you want it changed to two types of required, I think we should do it as another PR after this one is merged.